### PR TITLE
Add Sham-Needleamn GrainBoudnary cavitation model

### LINKED ISSUE #43: Add the Sham-Needleman GB Cavitation Model
as per title this PR will add the Sham-Needleman grain boundary cavitation model and all the tools required to make it work

### PR DESCRIPTION
**Pull request by arovinelli**
As per title, add the Sham Needleamn GB cavitation model
closes #43

### CODE REVIEW

File: src/materials/czm/GBCavitation/ShamNeedlemanEquation.C [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+    dfaldx[0] = 24. * L * a / std::pow(2 * a + 3 * L, 3.);
+  }
+  return dfaldx;
+}
+
+double V_dot::qFun(const bool implicit) {
+  // const double p = 30;
+  const double faL = faLFun(implicit);
+  const double fab = fabFun(implicit);
+  // const double f = std::pow(
```
Comment (reverendbedford) [inline src/materials/czm/GBCavitation/ShamNeedlemanEquation.C:124]: Commented out

File: src/materials/czm/GBCavitation/ShamNeedlemanEquation.C [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+  } else {
+    f = fab;
+    dqdx = dfabFundX(implicit);
+  }
+
+  const double dqdf = -2. * f - 2. / f + 4.;
+  for (uint i = 0; i < _n_vars; i++)
+    dqdx[i] *= dqdf;
+
+  // const double p = 30;
```
Comment (reverendbedford) [inline src/materials/czm/GBCavitation/ShamNeedlemanEquation.C:157]: Commented out

File: src/materials/czm/GBCavitation/ShamNeedlemanEquation.C [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+      _a0(a0), _growth_on(growth_on) {}
+
+double a_res::computedRate(const bool implicit) const {
+
+  // a_dot = V_dot/(4*pi*a^2*h)
+  double a_dot = 0;
+
+  if (_growth_on) {
+    // const double udot = _sysparams.getValue("udot_N");
+    // if (udot > 0 || (udot < 0 && _sys_vars.getValueOld("a") > _a0)) {
```
Comment (reverendbedford) [inline src/materials/czm/GBCavitation/ShamNeedlemanEquation.C:407]: Commented out

File: src/materials/czm/GBCavitation/ShamNeedlemanEquation.C [hunk 4]
```diff
@@ -0,0 +1,823 @@
+#pragma once
```
Comment (reverendbedford) [inline src/materials/czm/GBCavitation/ShamNeedlemanEquation.C:1]: Please add a glossary mapping the short variable names to what they are in the header (along with comments on what each function is).

File: src/utils/miconoss/Newton.C [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+               const uint _max_iter, const miconossmath::normtype normtype)
+    : _nlsys(_nlsys), _sys_vars(sys_vars), _tolerance(tolerance),
+      _max_iter(_max_iter), _n_eq(_nlsys->getNx()),
+      _n_lm(_nlsys->getNmultipliers()), _n_total(_n_eq + _n_lm), _dx(_n_total),
+      _normtype(normtype) {}
+
+int Newton::computeNewtonStep(const vecD &R, const matrixD &J) {
+  int ierr = miconossmath::solveAxb(J, R, _n_total, _dx);
+
+  for (uint i = 0; i < _n_total; i++)
```
Comment (reverendbedford) [inline src/utils/miconoss/Newton.C:13]: Uhh?  What?  Doesn't this zero out dx?

File: src/utils/miconoss/Newton.C [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  double err = 1e6;
+  int ierr = 0;
+  uint it = 0;
+  if (auto_scale_equation)
+    _nlsys->updateEquationScaling();
+
+  _nlsys->updateEquationConstants();
+  vecD R = _nlsys->assembleR(lm);
+  ierr = miconossmath::norm(R, _normtype, err);
+  if (ierr == 0) {
```
Comment (reverendbedford) [inline src/utils/miconoss/Newton.C:36]: Get rid of commented out statements or hide behind a debug flag.

**Review by reverendbedford — CHANGES_REQUESTED** [review]
1. For the "add comments" comments (mostly in the micronoss files) I'm not asking for exhaustive documentation, but please note what each class does and provide a description of the user-facing methods so that hopefully we can continue to use this in the future.
2. I don't see a complete CPFEM + GB example in the `examples/` subdirectory.
3. Variety of specific comments in the review.

File: src/utils/miconoss/Newton.C (hunk 5 above)
Comment (arovinelli) [reply src/utils/miconoss/Newton.C:13]: nope just inverting the sign

File: include/materials/czm/GBCavitation/GBCavitation.h [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  /// non linear solver paramters
+  ///@{
+  const unsigned int _max_time_cut;
+  const unsigned int _max_nonlinear_iter;
+  const Real _nl_residual_abs_tol;
+  const bool _force_substep;
+  ///@}
+
+  const GBCavitationBoundaryPropertyUO *_GBCavitationBoundaryPropertyUO;
+  /* INTERFACE PARAMETERS */
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/materials/czm/GBCavitation/GBCavitation.h:151]: there from the time when everything was a beautiful constant world... removed

File: include/materials/czm/GBCavitation/GBCavitation.h [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "CZMComputeLocalTractionIncrementalBase.h"
+#include "GBCavitationBoundaryPropertyUO.h"
+#include "ShamNeedlemanEquation.h"
+/**
+ * Implementation of teh grain boundary cavitation model proposed by Sham
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/materials/czm/GBCavitation/GBCavitation.h:16]: done

File: include/materials/czm/GBCavitation/ShamNeedlemanEquation.h [hunk 9]
```diff
@@ -0,0 +1,195 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/materials/czm/GBCavitation/ShamNeedlemanEquation.h:1]: done

File: include/materials/czm/GBCavitation/ShamNeedlemanEquation.h [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+public:
+  a_res(const unsigned int eq_index, NLSystemVars &sysvars,
+        NLSystemParameters &sysparams, NLPreEquationEvalautionCalc &pre_eval,
+        const double h, const double a0, const double theta = 0,
+        const bool growth_on = true);
+
+  double computedRate(const bool implicit) const override;
+  vecD DComputedRatetDx(const bool implicit) const override;
+  vecD DComputedRatetDP(const bool implicit) const override;
+  // double equationScalingRule() const override;
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/materials/czm/GBCavitation/ShamNeedlemanEquation.h:68]: removed

File: include/materials/czm/GBCavitation/ShamNeedlemanEquation.h [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+public:
+  TN_res(const unsigned int eq_index, NLSystemVars &sysvars,
+         NLSystemParameters &sysparams, NLPreEquationEvalautionCalc &pre_eval,
+         const double thickness, const double E_interface, const double P_mt,
+         const double _P_thickness, const double theta = 0);
+
+  double computedRate(const bool implicit) const override;
+  vecD DComputedRatetDx(const bool implicit) const override;
+  vecD DComputedRatetDP(const bool implicit) const override;
+  // double equationScalingRule() const override;
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/materials/czm/GBCavitation/ShamNeedlemanEquation.h:108]: removed

File: include/materials/czm/GBCavitation/ShamNeedlemanEquation.h [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+         NLSystemParameters &sysparams, NLPreEquationEvalautionCalc &pre_eval,
+         const uint shear_index, const double thickness,
+         const double eta_sliding, const double G_interface,
+         const double theta = 0);
+
+public:
+  double computedRate(const bool implicit) const override;
+  vecD DComputedRatetDx(const bool implicit) const override;
+  vecD DComputedRatetDP(const bool implicit) const override;
+  // double equationScalingRule() const override;
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/materials/czm/GBCavitation/ShamNeedlemanEquation.h:140]: removed

File: include/userobjects/GBCavitationBoundaryPropertyUO.h [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "InterfaceUserObject.h"
+
+/**
+ *  An interface userobject reading boundary based materail property nad storing
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/userobjects/GBCavitationBoundaryPropertyUO.h:15]: done

File: include/userobjects/GBCavitationBoundaryPropertyUO.h [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+ * them for alter use
+ */
+class GBCavitationBoundaryPropertyUO : public InterfaceUserObject {
+public:
+  static InputParameters validParams();
+  GBCavitationBoundaryPropertyUO(const InputParameters &parameters);
+  virtual ~GBCavitationBoundaryPropertyUO(){};
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/userobjects/GBCavitationBoundaryPropertyUO.h]: removed

File: include/userobjects/GBCavitationBoundaryPropertyUO.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/userobjects/GBCavitationBoundaryPropertyUO.h]: yes, updated

File: src/materials/czm/GBCavitation/GBCavitation.C [hunk 15]
```diff
@@ -0,0 +1,705 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "GBCavitation.h"
+#include "NLSystem.h"
+registerMooseObject("DeerApp", GBCavitation);
+
+InputParameters GBCavitation::validParams() {
+  InputParameters params =
+      CZMComputeLocalTractionIncrementalBase::validParams();
+  params.addClassDescription("Sham Needleman grain boundary cavitation model. "
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply src/materials/czm/GBCavitation/GBCavitation.C:17]: well these are Grade91 calibrated parameters, we can add a flag like `use_calibrated_Gr91_params` and if that is false I can force user to provide all material parameters

File: src/materials/czm/GBCavitation/GBCavitation.C [hunk 16]
```diff
@@ -0,0 +1,705 @@
...
+    _traction_at_failure[_qp] = _traction_at_failure_old[_qp];
+    _jump_at_failure[_qp] = _jump_at_failure_old[_qp];
+    _residual_life[_qp] = _residual_life_old[_qp];
+    _a[_qp] = _a_old[_qp];
+    _b[_qp] = _b_old[_qp];
+    _D[_qp] = _a[_qp] / _b[_qp];
+
+    tractionDecay();
+  } else {
+    computeAverageBulkPorperties();
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply src/materials/czm/GBCavitation/GBCavitation.C:245]: damn

File: src/materials/czm/GBCavitation/GBCavitation.C [hunk 17]
```diff
@@ -0,0 +1,705 @@
...
+      mooseError("psi_degree must be between 0 and 90 degrees");
+  }
+  if (_JxW_ref <= 0)
+    mooseError("the paremter h_ref must be positive");
+
+  if (_ndisp == 1)
+    mooseError("not implemented for 1D problems");
+}
+
+void GBCavitation::computeInterfaceTractionIncrementAndDerivatives() {
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply src/materials/czm/GBCavitation/GBCavitation.C:215]: Lol, I think I can do something here

File: src/materials/czm/GBCavitation/GBCavitation.C [hunk 18]
```diff
@@ -0,0 +1,705 @@
...
+            std::isfinite(_dinterface_traction_djump[_qp](i, j)),
+            "GBCavitation _dinterface_traction_djump[_qp](" +
+                std::to_string(i) + "," + std::to_string(j) +
+                ") is not finite: " +
+                std::to_string(_dinterface_traction_djump[_qp](i, j)) +
+                " increment i = " +
+                std::to_string(_interface_displacement_jump_inc[_qp](i)));
+      }
+    }
+    mooseAssert(std::isfinite(_dinterface_traction_djump[_qp].det()),
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply src/materials/czm/GBCavitation/GBCavitation.C:440]: done

File: src/materials/czm/GBCavitation/GBCavitation.C [hunk 19]
```diff
@@ -0,0 +1,705 @@
...
+                "determinatn of _dinterface_traction_djump is not finite");
+  }
+}
+
+/* TODO ADD innerpentration penalty for broken elements to add */
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply src/materials/czm/GBCavitation/GBCavitation.C:445]: this is done, so I removed the line

File: src/materials/czm/GBCavitation/ShamNeedlemanEquation.C (hunk 3 above)
Comment (arovinelli) [reply src/materials/czm/GBCavitation/ShamNeedlemanEquation.C:389]: removed

File: src/materials/czm/GBCavitation/ShamNeedlemanEquation.C (hunk 2 above)
Comment (arovinelli) [reply src/materials/czm/GBCavitation/ShamNeedlemanEquation.C]: removed

File: src/materials/czm/GBCavitation/ShamNeedlemanEquation.C (hunk 1 above)
Comment (arovinelli) [reply src/materials/czm/GBCavitation/ShamNeedlemanEquation.C:124]: removed

File: include/utils/miconoss/Newton.h [hunk 20]
```diff
@@ -0,0 +1,37 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/Newton.h:1]: done

File: include/utils/miconoss/miconosstype.h [hunk 21]
```diff
@@ -0,0 +1,44 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/miconosstype.h:1]: done

File: include/utils/miconoss/Equation.h [hunk 22]
```diff
@@ -0,0 +1,66 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/Equation.h:1]: done

File: include/utils/miconoss/InequalityConstraint.h [hunk 23]
```diff
@@ -0,0 +1,30 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/InequalityConstraint.h:1]: done

File: include/utils/miconoss/NLParameter.h [hunk 24]
```diff
@@ -0,0 +1,56 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/NLParameter.h:1]: done

File: include/utils/miconoss/NLPreEquationEvalautionCalc.h [hunk 25]
```diff
@@ -0,0 +1,47 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/NLPreEquationEvalautionCalc.h:1]: done

File: include/utils/miconoss/NLSystem.h [hunk 26]
```diff
@@ -0,0 +1,40 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/NLSystem.h:1]: done

File: include/utils/miconoss/NLVar.h [hunk 27]
```diff
@@ -0,0 +1,120 @@
+
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/utils/miconoss/NLVar.h:1]: done

File: src/utils/miconoss/Newton.C (hunk 6 above)
Comment (arovinelli) [reply src/utils/miconoss/Newton.C:36]: done

**Review by reverendbedford — APPROVED** [review]

**Pull request merged**